spacemeshos/smapp
# Fix killing of the stale Node process

There is no main issue, but it also closes #776 and #775.

When App is working for a while and then the User closes it, the App can not exit the `go-spacemesh` process somehow.

After discussion with @avive we decided to make a robust flow of closing the app and Node process:
1. Call the shutdown via GRPC API
2. Wait up to 20 seconds and if the process is still there — send `SIGINT` signal
3. Wait up to 20 seconds and if the process is still there — send `SIGKILL` signal to kill it finally

This approach also revealed some problems with the custom handling of closing App, which are fixed within the second commit. One thing that is left here untested and probably should be fixed in the future: handling of the condition branch if smeshing is turned on.

### ATTENTION! TESTS NEEDED!
> I've replaced `exec` (which runs the command inside a shell) with `spawn` (from [cross-spawn](https://github.com/moxystudio/node-cross-spawn) library), so this thing should be tested on each platform.
> **Please, test it on other platforms (Windows, Linux) or ask someone to do so. And write a report in this PR.**

### Things to test:
1. Is Node runs and works
2. Run Smapp and wait for 5+ minutes (layers started to sync), then try to close the App.
3. Run Smapp, then kill the `go-spacemesh` process (it should provoke an error in the Network screen) and press the "Restart Node" button on that screen.
